doc/error-handling/error_handling_design_spec.md

```diff
@@ -0,0 +1,429 @@
+
+1. **Error reporter class** - defines functions to write to ERROR_DB and publish the notifications.
+2. **Error listener class** - defines functions to register/de-register for notifications on a specific table. Applications need to specify the table name, operation such CREATE/DELETE/UPDATE, notification type (success/failure/both)  and callback function when instantiating the error listener class. Applications can instantiate multiple listeners to receive notifications from different tables.
+
+### 3.3.1 Event processing
+
+Below diagram shows how framework reports notifications to the registered listeners:
+
+![Framework high level design](images/error_handling_overview.png)
+
+The following is the sequence of events involved in reporting a failure notification from Syncd process to application:
+
+1. Syncd reports errors using a single notification channel to OrchAgent. Using a single notification channel ensures that order of the notifications is retained.
+2. After receiving the notification from Syncd, OrchAgent:
+   - Translates it from SAI data types to ERROR_DB data types
+   - Adds an entry in to error database. If the entry already exists, the corresponding failure code is updated.
+   - Publishes the notifications to respective error listeners.
+3. Error listener waits for the incoming notifications, filters them and invokes the application callback.
+
+
+
+The following is the sequence of events involved in reporting successful programming of an entry from Syncd process to application:
+
+1. Syncd reports the successful programming of an entry to OrchAgent using a notification channel.
+
+2. After receiving the notification from Syncd, OrchAgent:
+
+   - Translates it from SAI data types to ERROR_DB data types.
+
+   - Removes the entry from database. if present.
+   - Publishes the notifications to respective error listeners.
+
+3. Error listener waits for the incoming notifications, filters them and invokes the application callback.
+
+
+
+The following table describes how framework handles some of the notifications:
+
+| Previous Notification | Current Notification | Framework Action                                             |
+| --------------------- | -------------------- | ------------------------------------------------------------ |
+| Create failure        | Update failure       | Update the entry in the database and notify the registered applications |
+| Create failure        | Delete failure       | Remove the entry from database and notify the registered applications |
+| Create failure        | Update success       | Remove the entry from the database and notify the registered applications |
+| Create success        | Delete failure       | Add the entry to the database and notify the registered applications |
+| Delete failure        | Create success       | Remove the entry from the database and notify the registered applications |

```

Comment:
Does the applications get out of order notifications from feedback loop? How to handle in the case of it? Ex: User does create/delete/create and do you expect the error feedback come in order?

Reply:
Order of notifications will be preserved - because changes to APP_DB and ASIC_DB maintain the sequence. In case the same object fails multiple times, we need a unique transaction id to associate the operation and failure. To address this, we are looking at adding unique ID to each APP_DB operation and reporting the ID back as part of failure notification.